### ISSUE TITLE
How to add procedure(plsql) definition type to navigate window?

### ISSUE THREAD

**Issue by elkin5** [oracle/opengrok#2885]
I need to get the functions and procedure of a plsql file and then take them to an api rest that allows to be obtained

**Comment by vladak** [comment]
Not sure I understand the question fully - could you rephrase it as a list the steps perhaps ?

**Comment by elkin5** [comment]
The navigate window show this (variable, package, function):

![Capture](https://user-images.githubusercontent.com/4118198/62370482-dc73a900-b4f7-11e9-81f6-43e1cdc286bd.PNG)

i need add procedure to navigate window.

otherside, is it posible get this elements from the navigate window?

**Comment by vladak** [comment]
Okay, these are 2 different questions. Better handle one question in single Issue only.

The problem seems to be that `PROCEDURE` entries are not recognized as definitions. This is done by Universal ctags - just make you are using recent version. According to #2293 this should work in recent OpenGrok releases.

As for retrieving all definitions in a document(s) via RESTful API, I don't think it works currently - tried searching for a wildcard returns list of files however the definitions are represented as a single `...` entry for each file. Please file a separate issue to track this.

**Comment by vladak** [comment]
Also, it depends on the suffix of the filename: https://github.com/oracle/opengrok/blob/a391bead371079d671483425934a74aad8ec4344/opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/sql/PLSQLAnalyzerFactory.java#L33-L40

**Comment by vladak** [comment]
Anyow, some of the definitions are recognized so this is perhaps just a matter of adding more. Trying locally, I cannot even get ctags to recognize definitions of functions. Can you reduce the file to generic minimal contents ?

**Comment by elkin5** [comment]
The suffix (pkg) was added in the .java files.
I work on the latest version of opengrok.
The code viewer recognizes the procedure as definition but the navigation window does not show these definitions (procedure), how could I do it?

**Comment by vladak** [comment]
How do you know that the procedure is recognized as a definition ? If you do a search for definition, is it found ?

**Comment by vladak** [comment]
i.e. what are the contents of the `tags` file if you run `ctags --langmap=sql:+.pkg` on the file in question ? (assuming the file indeed has the `.pkg` suffix and `ctags` is the same ctags program as is used for running the Indexer)

**Comment by vladak** [comment]
By the way, you do not have to modify the source code of OpenGrok in order to add ctags suffixes: https://github.com/oracle/opengrok/wiki/Indexer-configuration#custom-ctags-configuration

**Comment by elkin5** [comment]
I search all definitions from one file.

![2-](https://user-images.githubusercontent.com/4118198/62377788-0d100e80-b509-11e9-8c04-acc0502d1fd0.png)

**Comment by elkin5** [comment]
it didn't work when trying to do it this way.
https://github.com/oracle/opengrok/wiki/Indexer-configuration#custom-ctags-configuration

**Comment by vladak** [comment]
I still do not know how the definitions come into existence, however maybe you can try playing a bit with https://github.com/oracle/opengrok/blob/81b586e651d9f2a40ab5a985d2b7080ae732dd83/opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/JFlexXrefUtils.java#L201-L203 , maybe adding `"procedure"` there will help.

**Comment by elkin5** [comment]
Ok, thank you!

**Comment by vladak** [comment]
Of course, this will need reindex as the `xref` file needs to be recreated using this new code.

**Comment by elkin5** [comment]
it don't work

**Comment by elkin5** [comment]
i will keep looking in the code

**Comment by vladak** [comment]
it would really help if you can publish a file somewhere that exhibits the problem.

**Comment by elkin5** [comment]
### Description problem
**Opengrok can search the definitions for diferents lenguajes (C#, java, pl/sql) and this definitions are used in diferents elements of the code, but example:**

**1. The navigate window:** This element show an abstract from all definitions content in the file.
### Problem 1
**For languaje pl/sql does not have in consider the procedure definitions what equal to functions void in java or C#**

![issueOG1_](https://user-images.githubusercontent.com/4118198/62986591-e94ea180-be01-11e9-8e2c-fcca181f90d4.png)

**2. The cross reference:** This allows to see in which method another method is used (image with search on java and search in pl/sql)
### Problem 2 
**For languaje pl/sql does not can show the method when is call the search method method**

**Search in .java file**

![issueOG2_](https://user-images.githubusercontent.com/4118198/62986642-27e45c00-be02-11e9-8488-cf96fdea6f39.png)

**Search in pl/sql (.pkg) file**

![issueOG3_](https://user-images.githubusercontent.com/4118198/62986717-7db90400-be02-11e9-9569-627902210fb9.png)

**3. The code navigation:** This allows navigating to method definitions within the code viewer.
### Problem 3
**For languaje pl/sql some methods are not recognized correctly**

![nn](https://user-images.githubusercontent.com/4118198/62986907-2d8e7180-be03-11e9-94b6-bf425ce8fb58.gif)

**Comment by idodeclare** [comment]
@elkin5 , see `JFlexXrefUtils.writeSymbolTable()` and `XrefStyle.DEFINITION_STYLES`.

The symbol data used for the Navigate Window is written (in the form of a JavaScript function actually) when an xref is generated. If you view source on an xref page, you'll see e.g.:

```js
function get_sym_list(){return [["Variable","xv",[["foo",16] ...
```

`"procedure"` is not in `XrefStyle.DEFINITION_STYLES`, but it could be added.

**Comment by elkin5** [comment]
**The initial question of this thread has been solved, therefore of the 3 problems presented**

### Problem 1: solved. 
> @elkin5 , see `JFlexXrefUtils.writeSymbolTable()` and `XrefStyle.DEFINITION_STYLES`.
> 
> The symbol data used for the Navigate Window is written (in the form of a JavaScript function actually) when an xref is generated. If you view source on an xref page, you'll see e.g.:
> 
> ```js
> function get_sym_list(){return [["Variable","xv",[["foo",16] ...
> ```
> 
> `"procedure"` is not in `XrefStyle.DEFINITION_STYLES`, but it could be added.

### Problem 2:  Did it create thread with this problem and i close this thread?

### Problem 3:  Did it create thread with this problem and i close this thread?

**Comment by vladak** [comment]
I think it is worth changing this in the OpenGrok source code. Keeping changes locally is no fun. Feel free to submit pull request with the changes.
